File: src/classes/renderer.cpp
#include "renderer.hpp"
#include "particle.hpp"
#include "../utils/texture_utils.hpp"
#include "plane.hpp"
#include "camera.hpp"
#include "molecule.hpp"
#include "container.hpp"
#include <glm/gtc/type_ptr.hpp>
#include <glew.h>
#include <fstream>
#include <sstream>
#include <string>
#include <iostream>
#include <vector>
#include <memory>
#include "mesh.hpp"
#include <omp.h>

void checkError(GLuint shaderProgram, const std::string& type) {
    if (shaderProgram == 0) {
        std::cerr << "Failed to create shader program of type " << type << std::endl;
        exit(EXIT_FAILURE);
    }
};

Renderer::Renderer() {



    // Create a shader program
    shaderProgram = createShaderProgram("shaders/vertexShader.glsl", "shaders/geometryShader.glsl", "shaders/fragmentShader.glsl");
    checkError(shaderProgram, "particle");

    floorShaderProgram = createShaderProgram("shaders/floorVertexShader.glsl", "shaders/floorFragmentShader.glsl");
    checkError(floorShaderProgram, "floor");

    modelShaderProgram = createShaderProgram("shaders/modelVertexShader.glsl", "shaders/modelFragmentShader.glsl");
    checkError(modelShaderProgram, "model");

    containerShaderProgram = createShaderProgram("shaders/containerVertexShader.glsl", "shaders/containerFragmentShader.glsl");
    checkError(containerShaderProgram, "container");

    moleculeLinksShaderProgram = createShaderProgram("shaders/modelOrientedVertexShader.glsl", "shaders/modelFragmentShader.glsl");
    checkError(moleculeLinksShaderProgram, "moleculeLinks");
}

void Renderer::draw(const Camera& camera, const std::vector<std::shared_ptr<Sphere>>& spheres) { // note : shared_ptr (*) meaning we take the pointer to the particle (and this allow polymorphism if we don't use the pointer we cannot use children classes) and the & meaning we take the reference to the particle
    std::vector<float> data;
    for (const auto& sphere : spheres) { 
        data.push_back(sphere->position.x);
        data.push_back(sphere->position.y);
        data.push_back(sphere->position.z);
        data.push_back(sphere->radius); // TODO fix this later to render the sphere properly
    }

    // Create a VBO and copy the data into it
    GLuint vbo;
    glGenBuffers(1, &vbo);
    glBindBuffer(GL_ARRAY_BUFFER, vbo);
    glBufferData(GL_ARRAY_BUFFER, data.size() * sizeof(float), data.data(), GL_STATIC_DRAW);

    // Use the shader program
    glUseProgram(shaderProgram);

    camera.loadMatricesIntoShader(shaderProgram);

    // Set up the vertex attribute pointers and draw the particles
    glEnableVertexAttribArray(0);
    glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 4 * sizeof(float), nullptr);
    glEnableVertexAttribArray(1);
    glVertexAttribPointer(1, 1, GL_FLOAT, GL_FALSE, 4 * sizeof(float), (void*)(3 * sizeof(float)));

    glDrawArrays(GL_POINTS, 0, spheres.size());

    // Clean up
    glDisableVertexAttribArray(0);
    glDisableVertexAttribArray(1);
    glBindBuffer(GL_ARRAY_BUFFER, 0);
    glDeleteBuffers(1, &vbo);
}

GLuint Renderer::loadAndCompileShader(const std::string& filename, GLenum shaderType) {
    // Load the shader source
    std::ifstream file("../" + filename);
    if (!file) {
        std::cerr << "Unable to open file " << filename << std::endl;
        return 0;
    }
    std::stringstream fileStream;
    fileStream << file.rdbuf();
    std::string shaderSrc = fileStream.str();
    const char* shaderSrcCStr = shaderSrc.c_str();

    // Create the shader
    GLuint shader = glCreateShader(shaderType);
    glShaderSource(shader, 1, &shaderSrcCStr, nullptr);
    glCompileShader(shader);

    // Check for errors
    GLint success;
    GLchar infoLog[512];
    glGetShaderiv(shader, GL_COMPILE_STATUS, &success);
    if (!success) {
        glGetShaderInfoLog(shader, 512, NULL, infoLog);
        std::cerr << "ERROR::SHADER::COMPILATION_FAILED\n" << infoLog << std::endl;
        return 0;
    }

    return shader;
}

GLuint Renderer::createShaderProgram(const std::string& vertexShaderFile, const std::string& fragmentShaderFile) {
    // Load the vertex shader
    GLuint vertexShader = loadAndCompileShader(vertexShaderFile, GL_VERTEX_SHADER);
    checkError(vertexShader, "vertex");

    // Load the fragment shader
    GLuint fragmentShader = loadAndCompileShader(fragmentShaderFile, GL_FRAGMENT_SHADER);
    checkError(fragmentShader, "fragment");

    // Create the shader program
    GLuint shaderProgram = glCreateProgram();
    glAttachShader(shaderProgram, vertexShader);
    glAttachShader(shaderProgram, fragmentShader);
    glLinkProgram(shaderProgram);

    // Clean up
    glDeleteShader(vertexShader);
    glDeleteShader(fragmentShader);

    return shaderProgram;
}

GLuint Renderer::createShaderProgram(const std::string& vertexShaderFile, const std::string& geometryShaderFile, const std::string& fragmentShaderFile) {
    // Load the vertex shader
    GLuint vertexShader = loadAndCompileShader(vertexShaderFile, GL_VERTEX_SHADER);
    checkError(vertexShader, "vertex");

    // Load the geometry shader
    GLuint geometryShader = loadAndCompileShader(geometryShaderFile, GL_GEOMETRY_SHADER);
    checkError(geometryShader, "geometry");

    // Load the fragment shader
    GLuint fragmentShader = loadAndCompileShader(fragmentShaderFile, GL_FRAGMENT_SHADER);
    checkError(fragmentShader, "fragment");

    // Create the shader program
    GLuint shaderProgram = glCreateProgram();
    glAttachShader(shaderProgram, vertexShader);
    glAttachShader(shaderProgram, geometryShader);
    glAttachShader(shaderProgram, fragmentShader);
    glLinkProgram(shaderProgram);

    // Clean up
    glDeleteShader(vertexShader);
    glDeleteShader(geometryShader);
    glDeleteShader(fragmentShader);

    return shaderProgram;
}

void Renderer::draw(const Camera& camera, const std::vector<std::shared_ptr<Sphere>>& spheres, Mesh& mesh) {
    std::vector<glm::vec3> positions;
    std::vector<glm::vec3> scales;

    positions.reserve(spheres.size());
    scales.reserve(spheres.size());
    
    for (const auto& s : spheres) {
        positions.emplace_back(s->position);
        scales.emplace_back(glm::vec3(s->radius));
    }

    // Use the shader program
    mesh.draw(modelShaderProgram, camera, positions, scales);
}

void Renderer::drawMoleculeLinks(const Camera& camera, const std::vector<std::shared_ptr<Molecule>>& molecules, Mesh& mesh) {
    std::vector<glm::vec3> positions;
    std::vector<glm::vec3> scales;
    std::vector<glm::vec3> rotations; // New vector for the rotation matrices

    for (const auto& molecule : molecules) {
        for (int i = 0; i < molecule->links.size(); i++) {
            std::shared_ptr<Sphere> s1 = molecule->links[i].first;
            std::shared_ptr<Sphere> s2 = molecule->links[i].second;
            glm::vec3 center = (s1->position + s2->position) / 2.0f;
            float distance = glm::length(s1->position - s2->position);
            positions.push_back(center);
            scales.push_back(glm::vec3(s1->radius / 2, distance / 2, s1->radius / 2));
            // scales.push_back(glm::vec3(0.25f));

            // Calculate the direction vector of the link
            glm::vec3 axis = glm::normalize(s2->position - s1->position);
            glm::vec3 modelUp = glm::vec3(0.0f, 1.0f, 0.0f);
            glm::vec3 modelX = glm::vec3(1.0f, 0.0f, 0.0f);
            glm::vec3 modelZ = glm::vec3(0.0f, 0.0f, 1.0f);

            // get the dot product of the axis and modelUp vectors
            float dotProduct = glm::dot(axis, modelUp);

            // clamp the value to the range [-1.0, 1.0] to avoid precision issues
            dotProduct = glm::clamp(dotProduct, -1.0f, 1.0f);

            // get the angle between the up axis and the axis
            float angleX = glm::degrees(acos(dotProduct));

            // get the cross product of the axis and modelUp vectors
            glm::vec3 crossProductX = glm::cross(axis, modelUp);

            // adjust the sign of the angle based on the direction of the cross product
            if (glm::dot(modelUp, crossProductX) < 0) {
                angleX = -angleX;
            }
            
            // get the projection of the axis on the xz plane
            glm::vec3 axisXZ = glm::normalize(axis - glm::dot(axis, modelUp) * modelUp);

            // get the cross product of the axisXZ and modelZ vectors
            glm::vec3 crossProduct = glm::cross(axisXZ, modelZ);

            // get the angle between the z axis and the projection of the axis on the xz plane
            float angleY = glm::degrees(atan2(glm::length(crossProduct), glm::dot(axisXZ, modelZ)));

            // check nan
            if (isnan(angleY)) {
                angleY = 0.0f;
            }

            if (glm::dot(modelUp, crossProduct) < 0) {
                angleY = -angleY;
            }

            angleY = angleY + 180.0f;

            glm::vec3 rotation = glm::vec3(angleX, angleY, 0.0f);
            rotations.push_back(rotation); // Add the rotation angles to the vector

        }
    }

    mesh.drawOriented(moleculeLinksShaderProgram, camera, positions, scales, rotations); // Pass the rotation matrices to the draw function
}

void Renderer::drawPlanes(const Camera& camera, const std::vector<std::shared_ptr<Plane>>& planes) {
    // Use the shader program
    glUseProgram(floorShaderProgram);

   camera.loadMatricesIntoShader(floorShaderProgram);

    for (const auto& plane : planes) {
        // Bind the floor texture
        glActiveTexture(GL_TEXTURE0);
        glBindTexture(GL_TEXTURE_2D, plane->getTexture());

        // Set the texture uniform
        glUniform1i(glGetUniformLocation(floorShaderProgram, "texture1"), 0);

        // Bind the floor VAO and draw the floor
        glBindVertexArray(plane->getVao());
        glDrawArrays(GL_TRIANGLES, 0, 6);

        // Unbind the VAO and the texture
        glBindVertexArray(0);
        glBindTexture(GL_TEXTURE_2D, 0);
    }
}

void Renderer::drawContainer(const Camera& camera, const std::vector<std::shared_ptr<Container>>& containers, Mesh& mesh) {
    std::vector<glm::vec3> positions;
    std::vector<glm::vec3> scales;

    for (const auto& container : containers) {
        positions.push_back(container->position);
        scales.push_back(container->size);
    }

    // Use the shader program
    mesh.draw(containerShaderProgram, camera, positions, scales);
}

File: rust/src/classes/renderer.rs
use glium::{Program, Display, VertexBuffer, Surface, Frame};
use glium::index::NoIndices;
use glium::program::ProgramCreationError;
use glium::uniforms::UniformsStorage;
use glium::implement_vertex;
use glium::draw_parameters::{DrawParameters};
use glutin::surface::{WindowSurface};
use std::fs::File;
use std::io;
use std::io::Read;
use std::path::Path;
use std::error::Error;
use std::sync::{Arc, RwLock};
use nalgebra::{Vector3};

use crate::config::{GLuint, CAMERA_ASPECT_RATIO, CAMERA_FOV, CAMERA_FAR, CAMERA_NEAR};
use crate::classes::particles::sphere::{Sphere};
use crate::classes::particle::{ParticleTrait};
use crate::classes::molecule::{Molecule};
use crate::classes::container::{ContainerTrait, Container};
use crate::classes::camera::{Camera};
use crate::classes::mesh::{Mesh, SubVboVertexBuffer};

// TODO do the renderer after

#[derive(Copy, Clone)]
struct SphereVertexBuffer {
    position: [f32; 3],
    radius: f32,
}
implement_vertex!(SphereVertexBuffer, position, radius);

#[derive(Copy, Clone)]
pub struct InstancePosVertex {
    pub instancePos: [f32; 3],
}
implement_vertex!(InstancePosVertex, instancePos);

#[derive(Copy, Clone)]
pub struct InstanceScaleVertex {
    pub instanceScale: [f32; 3],
}
implement_vertex!(InstanceScaleVertex, instanceScale);

#[derive(Copy, Clone)]
pub struct InstanceRotationVertex {
    pub instanceRotation: [f32; 3],
}
implement_vertex!(InstanceRotationVertex, instanceRotation);

pub struct Renderer {
    shader_program: Program,
    floor_shader_program: Program,
    // floor_vao: GLuint,
    // floor_vbo: GLuint,
    // floor_texture: GLuint,
    model_shader_program: Program,
    container_shader_program: Program,
    molecule_links_shader_program: Program,
}

impl Renderer {
    pub fn new(display: &Display<WindowSurface>) -> Result<Renderer, Box<dyn Error>> {
        let shader_program = Renderer::create_shader_program_with_geometry(display, "../shaders/vertexShader.glsl", "../shaders/geometryShader.glsl", "../shaders/fragmentShader.glsl")?;
        let floor_shader_program = Renderer::create_shader_program(display, "../shaders/floorVertexShader.glsl", "../shaders/floorFragmentShader.glsl")?;
        let model_shader_program = Renderer::create_shader_program(display, "../shaders/modelVertexShader.glsl", "../shaders/modelFragmentShader.glsl")?;
        let container_shader_program = Renderer::create_shader_program(display, "../shaders/containerVertexShader.glsl", "../shaders/containerFragmentShader.glsl")?;
        let molecule_links_shader_program = Renderer::create_shader_program(display, "../shaders/modelOrientedVertexShader.glsl", "../shaders/modelFragmentShader.glsl")?;

        Ok(Renderer {
            shader_program,
            floor_shader_program,
            model_shader_program,
            container_shader_program,
            molecule_links_shader_program,
        })
    }

    fn load_and_compile_shader<P: AsRef<Path>>(path: P) -> Result<String, Box<dyn Error>> {
        let mut file = match File::open(&path)  {
            Ok(file) => file,
            Err(e) => {
                eprintln!("Failed to open file: {}", path.as_ref().display());
                return Err(Box::new(e));
            }
        };
        let mut src = String::new();
        file.read_to_string(&mut src)?;
        Ok(src)
    }

    fn create_shader_program(display: &Display<WindowSurface>, vertex_shader_file: &str, fragment_shader_file: &str) -> Result<Program, ProgramCreationError> {
        let vertex_shader_src = Renderer::load_and_compile_shader(vertex_shader_file).unwrap();
        let fragment_shader_src = Renderer::load_and_compile_shader(fragment_shader_file).unwrap();

        Program::from_source(display, &vertex_shader_src, &fragment_shader_src, None)
    }

    fn create_shader_program_with_geometry(display: &Display<WindowSurface>, vertex_shader_file: &str, geometry_shader_file: &str, fragment_shader_file: &str) -> Result<Program, ProgramCreationError> {
        let vertex_shader_src = Renderer::load_and_compile_shader(vertex_shader_file).unwrap();
        let fragment_shader_src = Renderer::load_and_compile_shader(fragment_shader_file).unwrap();
        let geometry_shader_src = Renderer::load_and_compile_shader(geometry_shader_file).unwrap();

        Program::from_source(display, &vertex_shader_src, &fragment_shader_src, Some(&geometry_shader_src))
    }

    pub fn draw(&self, display: &Display<WindowSurface>, target: &mut Frame, camera: &Camera, spheres: &Vec<Arc<RwLock<Sphere>>>, mesh: &mut Mesh, params: &DrawParameters) {
        let mut positions = Vec::new();
        let mut scales = Vec::new();

        for sphere in spheres {
            let sphere = sphere.write().unwrap();
            positions.push(InstancePosVertex { instancePos: sphere.particle.position.into() });
            scales.push(InstanceScaleVertex { instanceScale: [sphere.radius, sphere.radius, sphere.radius] });
        }
        
        mesh.draw(&self.model_shader_program, camera, &positions, &scales, display, target, params);
    }

    pub fn draw_container(&self, display: &Display<WindowSurface>, target: &mut Frame, camera: &Camera, containers: &Vec<Arc<RwLock<dyn ContainerTrait>>>, mesh: &mut Mesh, params: &DrawParameters) {
        let mut positions = Vec::new();
        let mut scales = Vec::new();

        for container in containers {
            let container = container.write().unwrap();
            positions.push(InstancePosVertex { instancePos: container.get_position().into() });
            scales.push(InstanceScaleVertex { instanceScale: [container.get_size().x, container.get_size().y, container.get_size().z] });
        }
        
        mesh.draw(&self.container_shader_program, camera, &positions, &scales, display, target, params);
    }

    pub fn draw_molecule_links(&self, display: &Display<WindowSurface>, target: &mut Frame, camera: &Camera, molecules: &Vec<Arc<RwLock<Molecule>>>, mesh: &mut Mesh, params: &DrawParameters) {
        let mut positions = Vec::new();
        let mut scales = Vec::new();
        let mut rotations = Vec::new();

        for molecule in molecules {
            let molecule = molecule.read().unwrap();
            for link in &molecule.links {
                let s1 = link.0.read().unwrap();
                let s2 = link.1.read().unwrap();
                let center = (s1.get_position() + s2.get_position()) / 2.0;
                let distance = (s1.get_position() - s2.get_position()).norm();
                positions.push(InstancePosVertex { instancePos: center.into() });
                let current_scale = Vector3::new(s1.radius / 2.0, distance / 2.0, s1.radius / 2.0);
                scales.push(InstanceScaleVertex { instanceScale: current_scale.into() });

                let axis = (s2.get_position() - s1.get_position()).normalize();
                let model_up = Vector3::new(0.0, 1.0, 0.0);
                let model_z = Vector3::new(0.0, 0.0, 1.0);

                let dot_product = axis.dot(&model_up).clamp(-1.0, 1.0);
                let angle_x = dot_product.acos().to_degrees();
                let cross_product_x = axis.cross(&model_up);
                let angle_x = if model_up.dot(&cross_product_x) < 0.0 {
                    -angle_x
                } else {
                    angle_x
                };

                let axis_xz = (axis - model_up * axis.dot(&model_up)).normalize();
                let cross_product = axis_xz.cross(&model_z);
                let angle_y = cross_product.norm().atan2(axis_xz.dot(&model_z)).to_degrees();
                let angle_y = if angle_y.is_nan() { 0.0 } else { angle_y };
                let angle_y = if model_up.dot(&cross_product) < 0.0 {
                    -angle_y
                } else {
                    angle_y
                } + 180.0;

                let rotation = Vector3::new(angle_x, angle_y, 0.0);
                rotations.push(InstanceRotationVertex { instanceRotation: rotation.into() });
            }
        }

        mesh.draw_oriented(&self.molecule_links_shader_program, camera, &positions, &scales, &rotations, display, target, params);
    }
    
}